AnkiDroid/src/main/java/com/ichi2/anki/AndroidTtsPlayer.kt

```
@@ -0,0 +1,139 @@
+                        }
+                    }
+
+                    override fun onError(utteranceId: String?, errorCode: Int) {
+                        // TODO: Copied from ReadText
+                        Timber.v("Android TTS failed: %s (%d). Check logcat for error. Indicates a problem with Android TTS engine.", errorToDeveloperString(errorCode), errorCode)
+                        val helpUrl = Uri.parse(context.getString(R.string.link_faq_tts))
+                        context.mayOpenUrl(helpUrl)
+                        // TODO: We can do better in this UI now we have a reason for failure
+                        context.showSnackbar(R.string.no_tts_available_message) {
+                            setAction(R.string.help) { ReadText.openTtsHelpUrl(helpUrl) }
+                        }
+                    }
+                })
+            }
+            // optimisation: avoid the overhead of waiting for a completed job
+            createTtsJob = null
+        }
+    }
+
+    override fun onDestroy(owner: LifecycleOwner) {
+        super.onDestroy(owner)
+        tts?.stop()
+        tts?.shutdown()
+    }
+
+    override fun get_available_voices(): List<TtsVoice> {
+        return TtsVoices.allTtsVoices().toList()
+    }
+
+    override suspend fun play(tag: AvTag) {
+        if (tag !is TTSTag) {
+            Timber.w("Expected TTS Tag, got %s", tag)
+            return
+        }
+        val match = voice_for_tag(tag)
+        if (match == null) {
+            Timber.w("could not find voice for %s", tag)
+            return
+        }
+
+        val voice = match.voice
+        if (voice !is AndroidTtsVoice) {
+            Timber.w("Invalid voice for %s", tag)
+            return
+        }
+
+        play(tag, voice)
+    }
+
+    private suspend fun play(tag: TTSTag, voice: AndroidTtsVoice) {
+        val tts = requireTts().also {
+            it.voice = voice.voice
+            it.setSpeechRate(tag.speed)
```

Errors are ignored here